rfd/0169-auto-updates-linux-agents.md

```diff
@@ -0,0 +1,408 @@
+When restarting the agent during a downgrade, `systemd stop/start` are used before/after the downgrade.
+
+Teleport CA certificate rotations will break rollbacks.
+In the future, this could be addressed with additional validation of the agent's client certificate issuer fingerprints.
+For now, rolling forward will allow recovery from a broken rollback.
+
+Given that rollbacks may fail, we must maintain the following invariants:
+1. Broken rollbacks can always be reverted by reversing the rollback exactly.
+2. Broken versions can always be reverted by rolling back and then skipping the broken version.
+
+When rolling forward, the backup of the newer version's `sqlite.db` is only restored if that exact version is the roll-forward version.
+Otherwise, the older, rollback version of `sqlite.db` is preserved (i.e., the newer version's backup is not used).
+This ensures that a version upgrade which broke the database can be recovered with a rollback and a new patch.
+It also ensures that a broken rollback is always recoverable by reversing the rollback.
+
+Example: Given v1, v2, v3 versions of Teleport, where v2 is broken:
+1. v1 -> v2 -> v1 -> v3 => DB from v1 is migrated directly to v3, avoiding v2 breakage.
+2. v1 -> v2 -> v1 -> v2 -> v3 => DB from v2 is recovered, in case v1 database no longer has a valid certificate.
+
+### Manual Workflow
+
+For use cases that fall outside of the functionality provided by `teleport-updater`, we provide an alternative manual workflow using the `/v1/webapi/find` endpoint.
+This workflow supports customers that cannot use the auto-update mechanism provided by `teleport-updater` because they use their own automation for updates (e.g., JamF or Ansible).
+
+Cluster administrators that want to self-manage agent updates will be
+able to get and watch for changes to agent versions which can then be
+used to trigger other integrations to update the installed version of agents.
+
+```shell
+$ tctl autoupdate watch
+{"agent_version": "1.0.0", "agent_edition": "enterprise", ... }
+{"agent_version": "1.0.1", "agent_edition": "enterprise", ... }
+{"agent_version": "2.0.0", "agent_edition": "enterprise", ... }
+[...]
+```
+
+```shell
+$ tctl autoupdate get
+{"agent_version": "2.0.0", "agent_edition": "enterprise", ... }
+```
+
+### Installers
+
+The following install scripts will be updated to install the latest updater and run `teleport-updater enable` with the proxy address:
+- [/api/types/installers/agentless-installer.sh.tmpl](https://github.com/gravitational/teleport/blob/d0a68fd82412b48cb54f664ae8500f625fb91e48/api/types/installers/agentless-installer.sh.tmpl)
+- [/api/types/installers/installer.sh.tmpl](https://github.com/gravitational/teleport/blob/d0a68fd82412b48cb54f664ae8500f625fb91e48/api/types/installers/installer.sh.tmpl)
+- [/lib/web/scripts/oneoff/oneoff.sh](https://github.com/gravitational/teleport/blob/d0a68fd82412b48cb54f664ae8500f625fb91e48/lib/web/scripts/oneoff/oneoff.sh)
```

Comment:
The oneoff script is currently used by Discovery flows to configure AWS so that teleport can access AWS resources.
I don't think we need to update this script to run `teleport-updater`, because it's meant to run a single teleport command and be done with it (it is not installed into the system).

Even more important, I think we can use it to bootstrap the updater/first teleport installation.
All it does is:
- lookup OS/Arch
- download teleport binary
- run a `teleport arg1 arg2` command
(We can update the script to run teleport-updater instead)

It has some good test coverage, when comparing to other scripts.

This allow us to completely ignore the complexities of different package managers.

Reply:
+1 for using `oneoff.sh` as a minimal loader for the updater go binary. We'll need some form of bash script to do the glue (updater download+exec) on a new machine.